code-studio/src/js/pd/workshop_dashboard/components/workshop_form.jsx

```diff
@@ -0,0 +1,567 @@
+var Button = require('react-bootstrap').Button;
+var ButtonToolbar = require('react-bootstrap').ButtonToolbar;
+var Alert = require('react-bootstrap').Alert;
+
+var styles = {
+  readOnlyInput: {
+    backgroundColor: 'inherit',
+    cursor: 'default',
+    border: 'none'
+  }
+};
+
+var WorkshopForm = React.createClass({
+  contextTypes: {
+    router: React.PropTypes.object.isRequired
+  },
+
+  propTypes: {
+    workshop: React.PropTypes.shape({
+      id: React.PropTypes.number.isRequired,
+      facilitators: React.PropTypes.array.isRequired,
+      location_name: React.PropTypes.string.isRequired,
+      location_address: React.PropTypes.string.isRequired,
+      capacity: React.PropTypes.number.isRequired,
+      workshop_type: React.PropTypes.string.isRequired,
+      course: React.PropTypes.string.isRequired,
+      subject: React.PropTypes.string,
+      notes: React.PropTypes.string,
+      sessions: React.PropTypes.array.isRequired
+    }),
+    onSaved: React.PropTypes.func,
+    readOnly: React.PropTypes.bool
+  },
+
+  getInitialState: function () {
+    let initialState = {
+      errors: [],
+      shouldValidate: false,
+      facilitators: [],
+      location_name: '',
+      location_address: '',
+      capacity: '',
+      workshop_type: '',
+      course: '',
+      subject: '',
+      notes:'',
+      sessions: [{placeholderId: '_0'}],
+      destroyedSessions: [],
+      availableFacilitators: [],
+      showSaveConfirmation: false
+    };
+
+    if (this.props.workshop) {
+      initialState = _.merge(initialState,
+        _.pick(this.props.workshop, [
+        'facilitators',
+        'location_name',
+        'location_address',
+        'capacity',
+        'workshop_type',
+        'course',
+        'subject',
+        'notes'
+        ])
+      );
+      initialState.sessions = this.prepareSessionsForForm(this.props.workshop.sessions);
+      this.loadAvailableFacilitators(this.props.workshop.course);
+    }
+    return initialState;
+  },
+
+  componentDidMount: function () {
+    this.enableAutocompleteLocation();
+  },
+
+  componentWillUnmount: function () {
+    if (this.isGoogleMapsLoaded()) {
+      google.maps.event.clearInstanceListeners(this.autocomplete);
+    }
+    if (this.saveRequest) {
+      this.saveRequest.abort();
+    }
+    if (this.loadWorkshopRequest) {
+      this.loadWorkshopRequest.abort();
+    }
+  },
+
+  componentWillReceiveProps: function (nextProps) {
+    if (nextProps.readOnly && !this.props.readOnly) {
+      this.setState(this.getInitialState());
+    }
+  },
+
+  componentDidUpdate: function () {
+    this.enableAutocompleteLocation();
+  },
+
+  loadAvailableFacilitators: function (course) {
+    this.loadWorkshopRequest = $.ajax({
+      method: "GET",
+      url: `/api/v1/pd/course_facilitators?course=${course}`,
+      dataType: "json"
+    }).done(function (data) {
+      this.setState(_.set(this.state, 'availableFacilitators', data));
```

Comment:
I don't think the _.set is necessary here? `this.setState` doesn't need a full state object very time. If you only give it one field, it will leave other fields untouched, i.e. I think you can get away this `this.setState({availableFacilitators: data})`